src/main/scala/org/ergoplatform/mining/ErgoMiner.scala

```diff
@@ -263,41 +288,60 @@ class ErgoMiner(ergoSettings: ErgoSettings,
           Future.failed(new Exception("Failed to create candidate")))(Future.successful)
         sender() ! candBlockFuture
       } else {
+        log.info("Generating new candidate requested by external miner")
         val readersR = (readersHolderRef ? GetReaders).mapTo[Readers]
         val candBlockFuture = readersR.flatMap {
           case Readers(h, s: UtxoStateReader, m, _) =>
-            Future.fromTry(publicKeyOpt match {
-              case Some(pk) =>
-                val ts = txsToInclude.filter { tx => inputsNotSpent(tx, s) }
-                createCandidate(pk, h, m, desiredUpdate, s, ts)
-                  .map(candidate => updateCandidate(candidate, pk, ts)) //returns external candidate
-              case None => Failure(new Exception("Candidate could not be generated: no public key available"))
-            })
+            Future.fromTry(generateCandidate(h, m, s, Some(txsToInclude)).map(_.externalVersion))
           case _ =>
             Future.failed(new Exception("Invalid readers state, mining is possible in UTXO mode only"))
         }
         sender() ! candBlockFuture
       }
 
-    case solution: AutolykosSolution =>
-      val result: Future[Unit] = candidateOpt.map { c =>
-        val newBlock = completeBlock(c.candidateBlock, solution)
-        powScheme.validate(newBlock.header).map(_ => newBlock)
-      } match {
-        case Some(Success(newBlock)) =>
-          sendToNodeView(newBlock)
-          Future.successful(())
-        case Some(Failure(exception)) =>
-          Future.failed(new Exception(s"Improper block mined: ${exception.getMessage}", exception))
-        case None =>
-          Future.failed(new Exception("Invalid miner state"))
+    // solution found externally (by e.g. GPU miner)
+    case preSolution: AutolykosSolution =>
+      // Inject node pk if it is not externally set (in Autolykos 2)
+      val solution = if (preSolution.pk.isInfinity) {
+        publicKeyOpt match {
+          case Some(pk) => AutolykosSolution(pk.value, preSolution.w, preSolution.n, preSolution.d)
+          case None => preSolution
+        }
+      } else {
+        preSolution
       }
+      log.info("Got solution: " + solution)
+      val result: Future[Unit] =
```

Comment:
It seems this Future is actually used like Try. Why not make is into Try?
Also sending Future as a message is not a good idea as it may leak effects on internal mutable state of the actor.